File: sorting_algorithms/quick_sort/quicksort.scala
object quicksort extends App {
    def swap(arr: Array[Int], i: Int, j: Int) : Unit = {
        val temp = arr(i)
        arr(i) = arr(j)
        arr(j) = temp
    }
    
    def partition(arr: Array[Int], start: Int, end: Int) : Int = {
        val pivot = arr(end)
        var i = start - 1
        (start until end).foreach(j => {
            if (arr(j) <= pivot) {
                i += 1
                swap(arr, i, j)
            } })
            
        swap(arr, i + 1, end)
        i + 1
    }
    
    def quicksort(arr: Array[Int], start: Int, end: Int) : Unit = {
        if (start < end) {
            val pivot = partition(arr, start, end)
            quicksort(arr, start, pivot - 1)
            quicksort(arr, pivot + 1, end)
        }
    }

    val arr = Array(5, 4, 3, 2, 1)
    quicksort(arr, 0, arr.length - 1)
    println(arr.mkString(" "))
}

File: sorting_algorithms/quick_sort/quicksort.ts
function partition(A : Array<number>, start : number, end : number) {
    let pivot : number = A[end]
    let i : number = start - 1
    for (let j = start; j < end; j++) {
        if (A[j] >= pivot) { // <= for ascending order, >= for descending order 
            i++
            [A[i], A[j]] = [A[j], A[i]]
        }
    }
    [A[i + 1], A[end]] = [A[end], A[i + 1]]
    return i + 1
}

function quicksort(A : Array<number>, start : number, end : number) {
    if (start < end) {
        let q : number = partition(A, start, end)
        quicksort(A, start, q - 1)
        console.log(end)
        quicksort(A, q + 1, end)
    }
}

let A : Array<number> = [5, 4, 3, 2, 1]
quicksort(A, 0, A.length - 1)
console.log(A)
